testtools/UART_interface/esp_arduino_uart_interface.py

```diff
@@ -0,0 +1,164 @@
+
+        # Special per opt handling:
+        if "exit" in message and first_read:
+            time.sleep(serial_settings.wait_for_flash)
+            output = ser.in_waiting
+            while output < 4:
+                time.sleep(1)
+                output = ser.in_waiting
+                print("%d bytes in waiting" % output)
+
+        if ser.readable():
+            output = ser.readline(ser.in_waiting)
+            output = output.decode(encoding='utf-8', errors='ignore')
+
+            check_firmware_errors(output)
+            check_sdk_errors(output)
+            self.check_sample_errors(output)
+            print(output)
+            try:
+                # File must exist to write to it
+                file.writelines(output)
+            except:
+                pass
+
+            return output
+        else:
+            try:
+                time.sleep(2)
+                ser.open()
+                self.serial_read(ser, message, file, first_read=True)
+            except:
+                return False
+
+    # Note: the buffer size on the mxchip appears to be 128 Bytes.
+    def write_read(self, ser, input_file, output_file):
+        if 'esp32' in serial_settings.device_type:
+            serial_settings.bits_to_cache = 800
+            serial_settings.baud_rate = 1000000
+        else:
+            serial_settings.bits_to_cache = 1600
+            serial_settings.baud_rate = 115200
+
+        if input_file:
+            # set wait between read/write
+            wait = (serial_settings.bits_to_cache/serial_settings.baud_rate)
+            with open(input_file) as fp:
+                # initialize input/output_file
+                line = fp.readline()
+                f = open(output_file, 'w+') # Output file
+                while line:
+                    # time.sleep(.1)
+                    # Print only instruction, not secret content
+                    if line.split():
+                        print("Sending %s" %line.split()[0])
+
+                    # Attempt to write to serial port
+                    if not self.serial_write(ser, line, f):
+                        print("Failed to write to serial port, please diagnose connection.")
+                        f.close()
+                        break
+                    time.sleep(1)
+
+                    # Attempt to read serial port
+                    output = self.serial_read(ser, line, f, first_read=True)
+
+                    while(output):
+                        time.sleep(wait)
+                        output = self.serial_read(ser, line, f)
+                    line = fp.readline()
+
+                # read any trailing output, save to file
+                while (ser.in_waiting):
+                    time.sleep(.2)
+                    output = self.serial_read(ser, line, f)
+
+                # forward failed callbacks to SDK_ERRORS
+                azure_test_firmware_errors.SDK_ERRORS += 5 - self.message_callbacks
```

Comment:
This: [https://github.com/Azure/azure-iot-arduino/blob/master/examples/iothub_ll_telemetry_sample/telemetry_sample.c.ino#L46](url)
I suppose it is a bit magic numbery.

Reply:
although that number is unlikely to change, can we have a variable at the top that defines that number and sets it to five? just so this number doesn't get lost in the fray and we know easily where to change it and what it changes. kind of like how in C we don't have magic numbers but macro's at the top of files.

Reply:
Alternatively I could add a printout to the sample to relay how many times it's printing out and consume that with this interface, but given that we control the code on both ends, it may be an over design.

Reply:
yea i think just a variable at the top should be okay for now. just enough that we can spot it if we need to change it.